src/runtime/mod.rs

```diff
@@ -18,6 +18,18 @@ use crate::{
     options::StreamAddress,
 };
 
+#[cfg(feature = "tokio-runtime")]
+lazy_static::lazy_static! {
+    static ref SYNC_RUNTIME: std::sync::Mutex<tokio::runtime::Runtime> = {
+        let runtime = tokio::runtime::Builder::new()
+            .threaded_scheduler()
```

Comment:
This defaults to using a thread per CPU core. I'm wondering if this is what we want, considering that the user might be using other sync wrappers on top of async implementations (e.g. trust-dns or reqwest). If each of those also starts a thread per CPU, would we be overloading it a bit? This feels like a bit of a prisoner's dilemma in terms of each library getting the CPU time it needs.

Reply:
1000 threads, or 1000 processes? There's a bit of a difference, since in this case all the threads could be shared by the same process (i.e. their Rust application). In the case of your ~1000 processes, the single process containing the application might be getting only 1/1000 of the CPU time and then splitting that among the N threads, rather than the N threads each getting 1/1000 of the CPU time.

Reply:
Threads. I have around 250 or so processes running.

Reply:
It looks like both [trust-dns](https://github.com/bluejekyll/trust-dns/blob/76a3776d8840b4915390f241cda94d1f12b0df73/crates/resolver/src/resolver.rs#L77-L79) and [reqwest](https://github.com/seanmonstar/reqwest/blob/71368104e360edfdba78fb7b8dd23b8274f19c57/src/blocking/client.rs#L696) use the basic scheduler; I feel like we should probably do the same unless we have a pretty strong reason not to.

Reply:
I would say that performance here is a pretty strong reason not to. We're doing a ton of concurrent heavy lifting, and constraining that to a single thread seems needlessly restrictive. Consider the 0.9.x driver: it spawns a monitoring thread per server plus a background thread per connection pool, which is a minimum of 2 threads regardless of deployment. 

From the tokio docs:
>Most applications should use the threaded scheduler, except in some niche use-cases, such as when running only a single thread is required.

I don't think our driver is a niche use-case or that running only on a single thread is required.

Another thing that might be worth thinking about is allowing the user to configure the amount of threads that we use. If they want to restrict it, then they can. That said, even if we go that route, I strongly believe that the default should not be 1.

Reply:
I'm not super convinced by the 0.9.x driver precedent because that needs to use multiple threads in order to achieve concurrency, whereas our new synchronous client won't be doing nearly as much "concurrent heavy-lifting" because it still will still be using async tasks under the hood; in fact, we test the async client with only a single thread specifically because it's how we expect users will be using it. I think the correct way to think about the runtime here is "how many threads do we need to be able to run the internals of the driver (which are async)", and leave users to spawn additional threads for their own concurrent usage as needed; they're still free to spawn their own threads and share the client across them, in which case they won't be blocking any of the other driver functionality.

Note that the tokio docs says most "applications"; the driver is a library, not an application, so it's a bit harder to define a single correct value for all applications that the driver will be used with. Moreover, despite their documentation stating that, the tokio itself makes the default runtime single-threaded unless users opt-in to a feature that isn't enabled by default. From testing it locally, it seems like we don't even need that feature enabled in the driver anymore; it was only added so that we could run our tests with multiple threads while we were still blocking in parts of the driver.

My preference right now is still to enable only the basic scheduler and not letting it be configurable yet (since adding a configuration later is not a breaking API change, whereas if we add it now and want to remove it later, that would be a breaking change). Since I expect that you don't agree with this, I think we should probably take this discussion offline and circle back here once we all agree on a final decision.

Reply:
Per offline discussion, we're making `sync` a separate feature that depends on `async-std` for now.